File: wrappers/go/registry.go
// Package iso10383 is the Go wrapper for the ISO 10383 MIC registry.
//
// Construct with LoadRegistry() for the bundled snapshot, or
// LoadRegistryFromFile(path) to read a snapshot from disk.
package iso10383

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
)

//go:embed data/iso10383.json
var bundled []byte

// MIC is one entry of the registry. Fields that the source file may
// leave empty are pointers so that JSON null maps to a nil pointer
// rather than the zero value.
type MIC struct {
	MIC                string  `json:"mic"`
	MICType            string  `json:"mic_type"`
	Status             string  `json:"status"`
	OperatingMIC       string  `json:"operating_mic"`
	MarketName         *string `json:"market_name"`
	LegalEntityName    *string `json:"legal_entity_name"`
	LEI                *string `json:"lei"`
	MarketCategory     *string `json:"market_category"`
	Acronym            *string `json:"acronym"`
	CountryCode        *string `json:"country_code"`
	City               *string `json:"city"`
	Website            *string `json:"website"`
	CreationDate       *string `json:"creation_date"`
	LastUpdateDate     *string `json:"last_update_date"`
	LastValidationDate *string `json:"last_validation_date"`
	ExpirationDate     *string `json:"expiration_date"`
	Note               *string `json:"note"`
}

// BrokenChain records a segment whose parent chain does not terminate
// at an operating MIC. The bundled snapshot has zero of these.
type BrokenChain struct {
	MIC          string   `json:"mic"`
	OperatingMIC string   `json:"operating_mic"`
	Reason       string   `json:"reason"`
	Chain        []string `json:"chain"`
}

// Counts mirrors meta.counts.
type Counts struct {
	Operating int `json:"operating"`
	Segment   int `json:"segment"`
	Active    int `json:"active"`
	Updated   int `json:"updated"`
	Expired   int `json:"expired"`
	Total     int `json:"total"`
}

// Meta mirrors the top-level meta object.
type Meta struct {
	Version        string        `json:"version"`
	Updated        string        `json:"updated"`
	SourceSnapshot string        `json:"source_snapshot"`
	SourceURL      string        `json:"source_url"`
	SourceHash     string        `json:"source_hash"`
	Counts         Counts        `json:"counts"`
	BrokenChains   []BrokenChain `json:"broken_chains"`
}

type raw struct {
	Meta Meta  `json:"meta"`
	MICs []MIC `json:"mics"`
}

// Registry is a read-only view over a MIC snapshot.
type Registry struct {
	meta  Meta
	mics  []MIC
	byMIC map[string]int
}

// LoadRegistry loads the bundled snapshot.
func LoadRegistry() (*Registry, error) {
	return LoadRegistryFromBytes(bundled)
}

// LoadRegistryFromFile loads a snapshot from disk.
func LoadRegistryFromFile(path string) (*Registry, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return LoadRegistryFromBytes(b)
}

// LoadRegistryFromBytes parses a snapshot from a byte slice.
func LoadRegistryFromBytes(b []byte) (*Registry, error) {
	var r raw
	if err := json.Unmarshal(b, &r); err != nil {
		return nil, fmt.Errorf("parse registry: %w", err)
	}
	if r.Meta.Version == "" {
		return nil, fmt.Errorf("snapshot missing meta.version")
	}
	by := make(map[string]int, len(r.MICs))
	for i, m := range r.MICs {
		by[m.MIC] = i
	}
	return &Registry{meta: r.Meta, mics: r.MICs, byMIC: by}, nil
}

// Meta returns the snapshot metadata.
func (r *Registry) Meta() Meta { return r.meta }

// Len is the number of entries.
func (r *Registry) Len() int { return len(r.mics) }

// IsEmpty reports whether the registry has no entries.
func (r *Registry) IsEmpty() bool { return len(r.mics) == 0 }

// All returns a defensive copy of every entry.
func (r *Registry) All() []MIC {
	out := make([]MIC, len(r.mics))
	copy(out, r.mics)
	return out
}

// ByMIC returns a copy of the entry for mic, or nil.
// The lookup is case-insensitive.
func (r *Registry) ByMIC(mic string) *MIC {
	i, ok := r.byMIC[strings.ToUpper(mic)]
	if !ok {
		return nil
	}
	m := r.mics[i]
	return &m
}

// Segments returns the direct children of an operating MIC.
//
// A grandchild whose parent is another segment is not listed here.
// Walk with OperatingMIC for multi-level chains.
func (r *Registry) Segments(mic string) []MIC {
	u := strings.ToUpper(mic)
	var out []MIC
	for _, m := range r.mics {
		if m.MICType == "SEGMENT" && m.OperatingMIC == u {
			out = append(out, m)
		}
	}
	return out
}

// OperatingMIC returns the operating_mic field of mic, or "" if the
// MIC is unknown. Operating MICs self-reference (D3).
func (r *Registry) OperatingMIC(mic string) string {
	if m := r.ByMIC(mic); m != nil {
		return m.OperatingMIC
	}
	return ""
}

// Expired returns expired entries, sorted by expiration_date ascending.
// Pass "" for since to return every expired entry.
func (r *Registry) Expired(since string) []MIC {
	var out []MIC
	for _, m := range r.mics {
		if m.Status != "EXPIRED" {
			continue
		}
		if since != "" {
			d := ""
			if m.ExpirationDate != nil {
				d = *m.ExpirationDate
			}
			if d < since {
				continue
			}
		}
		out = append(out, m)
	}
	sort.Slice(out, func(i, j int) bool {
		di, dj := "", ""
		if out[i].ExpirationDate != nil {
			di = *out[i].ExpirationDate
		}
		if out[j].ExpirationDate != nil {
			dj = *out[j].ExpirationDate
		}
		return di < dj
	})
	return out
}

// ByCountry returns entries whose ISO 3166 alpha-2 country matches code.
func (r *Registry) ByCountry(code string) []MIC {
	u := strings.ToUpper(code)
	var out []MIC
	for _, m := range r.mics {
		if m.CountryCode != nil && *m.CountryCode == u {
			out = append(out, m)
		}
	}
	return out
}

// ByStatus returns entries whose status matches.
func (r *Registry) ByStatus(status string) []MIC {
	u := strings.ToUpper(status)
	var out []MIC
	for _, m := range r.mics {
		if m.Status == u {
			out = append(out, m)
		}
	}
	return out
}

// ByMICType returns entries whose mic_type matches.
func (r *Registry) ByMICType(micType string) []MIC {
	u := strings.ToUpper(micType)
	var out []MIC
	for _, m := range r.mics {
		if m.MICType == u {
			out = append(out, m)
		}
	}
	return out
}

// ByCategory returns entries whose market_category matches.
func (r *Registry) ByCategory(category string) []MIC {
	var out []MIC
	for _, m := range r.mics {
		if m.MarketCategory != nil && *m.MarketCategory == category {
			out = append(out, m)
		}
	}
	return out
}

// Search does a case-insensitive substring match on market_name and
// acronym.
func (r *Registry) Search(q string) []MIC {
	q = strings.ToLower(q)
	var out []MIC
	for _, m := range r.mics {
		name := ""
		if m.MarketName != nil {
			name = strings.ToLower(*m.MarketName)
		}
		acr := ""
		if m.Acronym != nil {
			acr = strings.ToLower(*m.Acronym)
		}
		if strings.Contains(name, q) || strings.Contains(acr, q) {
			out = append(out, m)
		}
	}
	return out
}

// Validate returns ok=true and an empty slice when every MIC exists.
// The second return value lists the MICs that were not found, in the
// order they were given.
func (r *Registry) Validate(mics []string) (bool, []string) {
	var missing []string
	for _, m := range mics {
		if _, ok := r.byMIC[strings.ToUpper(m)]; !ok {
			missing = append(missing, m)
		}
	}
	return len(missing) == 0, missing
}

File: wrappers/python/src/iso10383/registry.py
"""MICRegistry - read-only access to an iso10383 snapshot."""

from __future__ import annotations

import json
from dataclasses import dataclass
from importlib.resources import files
from pathlib import Path
from typing import Any, Iterable


def _bundled_path() -> Path:
    return Path(str(files("iso10383").joinpath("data/iso10383.json")))


@dataclass(frozen=True)
class MIC:
    mic: str
    mic_type: str
    status: str
    operating_mic: str
    market_name: str | None = None
    legal_entity_name: str | None = None
    lei: str | None = None
    market_category: str | None = None
    acronym: str | None = None
    country_code: str | None = None
    city: str | None = None
    website: str | None = None
    creation_date: str | None = None
    last_update_date: str | None = None
    last_validation_date: str | None = None
    expiration_date: str | None = None
    note: str | None = None

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "MIC":
        return cls(**{k: d.get(k) for k in cls.__dataclass_fields__})

    def as_dict(self) -> dict[str, Any]:
        return {k: getattr(self, k) for k in self.__dataclass_fields__}


class MICRegistry:
    """Load iso10383.json and answer MIC queries.

    Passing no source loads the bundled snapshot. Passing a path reads
    that file. Passing a parsed dict is also accepted.
    """

    def __init__(self, source: str | Path | dict | None = None):
        if source is None:
            data = json.loads(_bundled_path().read_text(encoding="utf-8"))
        elif isinstance(source, dict):
            data = source
        else:
            data = json.loads(Path(source).read_text(encoding="utf-8"))

        self._meta: dict[str, Any] = data["meta"]
        self._mics: list[dict[str, Any]] = data["mics"]
        self._by_mic: dict[str, dict[str, Any]] = {
            m["mic"]: m for m in self._mics
        }

    # --- metadata ---

    @property
    def version(self) -> str:
        return self._meta["version"]

    @property
    def updated(self) -> str:
        return self._meta["updated"]

    @property
    def source_snapshot(self) -> str:
        return self._meta["source_snapshot"]

    @property
    def source_hash(self) -> str:
        return self._meta["source_hash"]

    @property
    def counts(self) -> dict[str, int]:
        return dict(self._meta["counts"])

    @property
    def broken_chains(self) -> list[dict[str, Any]]:
        return list(self._meta.get("broken_chains", []))

    # --- queries ---

    def __len__(self) -> int:
        return len(self._mics)

    def all(self) -> list[MIC]:
        return [MIC.from_dict(m) for m in self._mics]

    def by_mic(self, mic: str) -> MIC | None:
        m = self._by_mic.get(mic.upper())
        return MIC.from_dict(m) if m else None

    def segments(self, mic: str) -> list[MIC]:
        """Direct children of an operating MIC.

        A grandchild whose parent is another segment is not listed
        here. Walk with `operating_mic` for multi-level chains.
        """
        u = mic.upper()
        return [MIC.from_dict(m) for m in self._mics
                if m["mic_type"] == "SEGMENT" and m["operating_mic"] == u]

    def operating_mic(self, mic: str) -> str | None:
        m = self._by_mic.get(mic.upper())
        return m["operating_mic"] if m else None

    def expired(self, since: str | None = None) -> list[MIC]:
        rows = [m for m in self._mics if m["status"] == "EXPIRED"]
        if since:
            rows = [m for m in rows
                    if (m.get("expiration_date") or "") >= since]
        rows.sort(key=lambda m: m.get("expiration_date") or "")
        return [MIC.from_dict(m) for m in rows]

    def by_country(self, code: str) -> list[MIC]:
        u = code.upper()
        return [MIC.from_dict(m) for m in self._mics
                if m.get("country_code") == u]

    def by_status(self, status: str) -> list[MIC]:
        u = status.upper()
        return [MIC.from_dict(m) for m in self._mics if m["status"] == u]

    def by_mic_type(self, mic_type: str) -> list[MIC]:
        u = mic_type.upper()
        return [MIC.from_dict(m) for m in self._mics if m["mic_type"] == u]

    def by_category(self, category: str) -> list[MIC]:
        return [MIC.from_dict(m) for m in self._mics
                if m.get("market_category") == category]

    def search(self, query: str) -> list[MIC]:
        q = query.lower()
        return [MIC.from_dict(m) for m in self._mics
                if q in (m.get("market_name") or "").lower()
                or q in (m.get("acronym") or "").lower()]

    def validate(self, mics: Iterable[str]) -> tuple[bool, list[str]]:
        missing = [m for m in mics if m not in self._by_mic]
        return (not missing, missing)
